src/docs/05-go-client/18-workflow-replay-shadowing.md

```diff
@@ -0,0 +1,176 @@
+
+```go
+// if workflow history has been loaded into memory
+err := replayer.ReplayWorkflowHistory(logger, history)
+
+// if workflow history is stored in a json file
+err = replayer.ReplayWorkflowHistoryFromJSONFile(logger, jsonFileName)
+
+// if workflow history is stored in a json file and you only want to replay part of it
+// NOTE: lastEventID can't be set arbitrarily. It must be the end of of a history events batch
+// when in doubt, set to the eventID of decisionTaskStarted events.
+err = replayer.ReplayPartialWorkflowHistoryFromJSONFile(logger, jsonFileName, lastEventID)
+
+// if you want to fetch workflow history directly from cadence server
+// please check the Worker Service page for how to create a cadence service client
+err = replayer.ReplayWorkflowExecution(ctx, cadenceServiceClient, logger, domain, execution)
+```
+
+#### Step 5: Check returned error
+
+If an error is returned from the replay method, it means there's a incompatible change in the workflow definition and the error message will contain more information regarding where the non-deterministic error happens.
+
+Note: currently an error will be returned if there're less than 3 events in the history. It is because first 3 events in the history has nothing to do with the workflow code, so Replayer can't tell if there's a incompatible change or not.
+
+### Sample Replay Test
+
+This sample is also available in our samples repo at [here](https://github.com/uber-common/cadence-samples/blob/master/cmd/samples/recipes/helloworld/replay_test.go#L39).
+
+```go
+func TestReplayWorkflowHistoryFromFile(t *testing.T) {
+	replayer := worker.NewWorkflowReplayer()
+	replayer.RegisterWorkflow(helloWorldWorkflow)
+	err := replayer.ReplayWorkflowHistoryFromJSONFile(zaptest.NewLogger(t), "helloworld.json")
+	require.NoError(t, err)
+}
+```
+
+## Workflow Shadower
+
+Workflow Replayer works well when verifying the compatibility against a small number of workflow histories. If there are lots of workflows in production need to be verified, dumping all histories manually clearly won't work. Directly fetching histories from cadence server might be a solution, but the time to replay all workflow histories might be too long for a test.
+
+Workflow Shadower is built on top of Workflow Replayer to address this problem. The basic idea of shadowing is: scan workflows based on the filters you defined, fetch history for each of workflow in the scan result from Cadence server and run the replay test. It can be run either as a test to serve local development purpose or as a workflow in your worker to continuously replay production workflows. 
+
+### Shadow Options
+
+Complete documentation on shadow options which includes default values, accepted values, etc. can be found [here](https://github.com/uber-go/cadence-client/blob/master/internal/workflow_shadower.go#L53). The following sections are just a brief description of each option.
+
+#### Scan Filters
```

Comment:
I'm not sure if vuepress supports h4's, if it does then can leave as-is. If it doesn't just change to a h3.

Reply:
I think it's used somewhere else for example here: https://cadenceworkflow.io/docs/go-client/activities/#heart-beating and it's working fine.